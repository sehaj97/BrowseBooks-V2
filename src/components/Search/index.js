import { useState, useEffect } from 'react';
import { BiSearch, BiCaretDown, BiCheck } from "react-icons/bi";

const DropDown = ({toggle, orderBy, onOrderByChange, sortBy, onSortByChange}) => {
    if(!toggle){
        return null
    }
    return (    <div className="origin-top-right absolute mt-2 w-56
    rounded-md shadow-lg bg-white ring-1 ring-black ring-opacity-5">
    <div className="py-1" role="menu" aria-orientation="vertical" aria-labelledby="options-menu">
      <div onClick={()=> {onSortByChange('title')}} tabIndex='0'
        className="px-4 py-2 text-sm text-gray-700 hover:bg-gray-100 hover:text-gray-900 flex justify-between cursor-pointer"
        role="menuitem">Book Name {(sortBy === 'title')&&<BiCheck />}</div>
      <div onClick={()=> {onSortByChange('first_publish_year')}} tabIndex='0'
        className="px-4 py-2 text-sm text-gray-700 hover:bg-gray-100 hover:text-gray-900 flex justify-between cursor-pointer"
        role="menuitem">Date {(sortBy === 'first_publish_year')&&<BiCheck />}</div>
      <div onClick={()=> {onOrderByChange('asc')}} tabIndex='0'
        className="px-4 py-2 text-sm text-gray-700 hover:bg-gray-100 hover:text-gray-900 flex justify-between cursor-pointer border-gray-1 border-t-2"
        role="menuitem">Asc {(orderBy === 'asc')&&<BiCheck />}</div>
      <div onClick={()=> {onOrderByChange('desc')}} tabIndex='0'
        className="px-4 py-2 text-sm text-gray-700 hover:bg-gray-100 hover:text-gray-900 flex justify-between cursor-pointer"
        role="menuitem">Desc {(orderBy === 'desc')&&<BiCheck />}</div>
    </div>
  </div>)
}

export const Search = ({setBooks,books,setErrorMessage,errorMessage, orderBy, onOrderByChange, sortBy, onSortByChange}) => {
  const [toggleSort, setToggleSort] = useState(false);
  const [query, setQuery] = useState('');
  const [search, setSearch] = useState('');

    function handleChange(e) {
        if (!e.target.value.length) {
            setErrorMessage(`${e.target.name} is required.`);
          } else {
            setErrorMessage('');
          }
          
        if (!errorMessage) {
          setSearch(e.target.value);
        }
      }
      
    function handleSubmit(e) {
        e.preventDefault();
        if (!search) {
          setErrorMessage(`input is required.`);
        } else {
          setErrorMessage('');
          setQuery(search);
        }
      }

  useEffect(() => {
    try {
      getData()
    } catch {
      setErrorMessage('something went wrong')
    }

    // we will use async/await to fetch this data
    async function getData() {
      const response = await fetch(`https://openlibrary.org/search.json?title=${query}&fields=title,author_name,isbn,first_publish_year`);
      const data = await response.json();
      if(data){
        // store the data into our books variable
        setBooks(data.docs) ;
      }
      
    }
  }, [query]);

    return(
        <div className="py-5 mx-2">
            <form onSubmit={handleSubmit} className="mt-1 flex rounded-md shadow-sm">
                <input type="text" name="query" id="query" onChange={handleChange} 
                className="pl-4 py-2 focus:ring-indigo-500 focus:border-indigo-500 w-full sm:text-sm border-gray-300" placeholder="Search" />
                
                    
                <button type="submit" 
                    className="px-4 py-2 bg-blue-400 border-2 border-blue-400 text-sm text-white hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2" id="options-menu" aria-haspopup="true" aria-expanded="true"><BiSearch/>
                </button>
                
            
                <div className="flex items-center">
                    <div>
                    <button type="button" onClick={()=>{setToggleSort(!toggleSort)}}
                        className="justify-center px-4 py-2 bg-teal-500 border-2 border-teal-500 text-sm text-white hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 flex items-center" id="options-menu" aria-haspopup="true" aria-expanded="true">
                        Sort By <BiCaretDown className="ml-2" />
                    </button>
                    <DropDown toggle={toggleSort}
                              orderBy={orderBy} 
                              onOrderByChange={(sortVal) => onOrderByChange(sortVal)} 
                              sortBy={sortBy} 
                              onSortByChange={(sortVal) => onSortByChange(sortVal)}/>
                    </div>
                </div>
            </form>
        </div>
    )
}