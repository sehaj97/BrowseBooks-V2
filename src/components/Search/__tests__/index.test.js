import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Search from '..';

afterEach(cleanup);

describe('Search component', () => {
  it('renders', () => {
    render(<Search />);
  });

  it('matches snapshot DOM node structure', () => {
    const { asFragment } = render(<Search />);
    expect(asFragment()).toMatchSnapshot();
  });

  })