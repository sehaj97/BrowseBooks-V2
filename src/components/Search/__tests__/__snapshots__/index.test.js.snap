// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component matches snapshot DOM node structure 1`] = `
<DocumentFragment>
  <div
    class="py-5 mx-2"
  >
    <form
      class="mt-1 flex rounded-md shadow-sm"
    >
      <input
        class="pl-4 py-2 focus:ring-indigo-500 focus:border-indigo-500 w-full sm:text-sm border-gray-300"
        id="query"
        name="query"
        placeholder="Search"
        type="text"
      />
      <button
        aria-expanded="true"
        aria-haspopup="true"
        class="px-4 py-2 bg-blue-400 border-2 border-blue-400 text-sm text-white hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2"
        id="options-menu"
        type="submit"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 18a7.952 7.952 0 0 0 4.897-1.688l4.396 4.396 1.414-1.414-4.396-4.396A7.952 7.952 0 0 0 18 10c0-4.411-3.589-8-8-8s-8 3.589-8 8 3.589 8 8 8zm0-14c3.309 0 6 2.691 6 6s-2.691 6-6 6-6-2.691-6-6 2.691-6 6-6z"
          />
        </svg>
      </button>
      <div
        class="flex items-center"
      >
        <div>
          <button
            aria-expanded="true"
            aria-haspopup="true"
            class="justify-center px-4 py-2 bg-teal-500 border-2 border-teal-500 text-sm text-white hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 flex items-center"
            id="options-menu"
            type="button"
          >
            Sort By 
            <svg
              class="ml-2"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m11.998 17 7-8h-14z"
              />
            </svg>
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
